code/modules/antagonists/traitor/datum_traitor.dm

```diff
@@ -153,33 +162,94 @@
 	employer = pick(possible_employers)
 	traitor_flavor = strings(TRAITOR_FLAVOR_FILE, employer)
 
+/// Generates a complete set of traitor objectives up to the traitor objective limit, including non-generic objectives such as martyr and hijack.
 /datum/antagonist/traitor/proc/forge_traitor_objectives()
 	objectives.Cut()
+	var/objective_count = 0
+
+	if((GLOB.joined_player_list.len >= HIJACK_MIN_PLAYERS) && prob(HIJACK_PROB))
+		is_hijacker = TRUE
+		objective_count++
 
-	var/list/kill_targets = list() //for blacklisting already-set targets
 	var/objective_limit = CONFIG_GET(number/traitor_objectives_amount)
-	for(var/i in 1 to objective_limit)
-		var/list/ai_targets = active_ais(z = 2) //For multiZ stations, this proc will include AIs on all station levels if the provided arg is 2.
-		ai_targets -= kill_targets
-		if(ai_targets.len)
-			var/diceroll = rand(1, (living_player_count() - 1)) //AI kill and crew kill objectives are different, but I want a rough equal chance for AIs to be targets. Maybe I'll refractor this someday
-			if(diceroll <= ai_targets.len)
-				var/datum/objective/task = new /datum/objective/destroy()
-				task.owner = owner
-				task.find_target(blacklist = kill_targets)
-				kill_targets += task.target
-				objectives += task
-				continue
-
-		var/datum/objective/task = new /datum/objective/assassinate()
-		task.owner = owner
-		task.find_target(blacklist = kill_targets)
-		kill_targets += task.target
-		objectives += task
-
-	var/datum/objective/escape/bye = new /datum/objective/escape()
-	objectives += bye
-	bye.owner = owner
+
+	// for(in...to) loops iterate inclusively, so to reach objective_limit we need to loop to objective_limit - 1
+	// This does not give them 1 fewer objectives than intended.
+	for(var/i in objective_count to objective_limit - 1)
+		objectives += forge_single_generic_objective()
+
+/**
+ * ## forge_ending_objective
+ *
+ * Forges the endgame objective and adds it to this datum's objective list.
+ */
+/datum/antagonist/traitor/proc/forge_ending_objective()
+	if(is_hijacker)
+		ending_objective = new /datum/objective/hijack
+		ending_objective.owner = owner
+		return
+
+	var/martyr_compatibility = TRUE
+
+	for(var/datum/objective/traitor_objective in objectives)
+		if(!traitor_objective.martyr_compatible)
+			martyr_compatibility = FALSE
+			break
+
+	if(martyr_compatibility && prob(MARTYR_PROB))
+		ending_objective = new /datum/objective/martyr
+		ending_objective.owner = owner
+		objectives += ending_objective
+		return
+
+	ending_objective = new /datum/objective/escape
+	ending_objective.owner = owner
+	objectives += ending_objective
+
+/// Forges a single escape objective and adds it to this datum's objective list.
+/datum/antagonist/traitor/proc/forge_escape_objective()
```

Comment:
What's this proc? It looks very similar to the one immediately above it but isn't called anywhere.

Reply:
not actually sure